### ISSUE TITLE
SparseDropout: Investigate up-scaling in training mode (like in regular dropout)

### ISSUE THREAD

**Issue by alopez** [RasaHQ/rasa#8954]
#### Current Status

*   The DenseForSparse layer realises dropout [like this](https://github.com/RasaHQ/rasa/blob/main/rasa/utils/tensorflow/layers.py#L82)
*   If the sparse tensor represents a sequence of tokens mapped to indices (with one index per token), the layer produces a sequence of corresponding dense embedding vectors, and hence the effect of the dropout is to just replace (in expectation) `rate*seq_len` many of these embedding vectors with padding (zero) vectors.
*   Now we can observe that:

1.  Even for short sentences, we might drop more words than we'd like. For example, given a sentence with 5 words and the standard drop probability of ~20%, we'll drop no word with a probability of ~30%, we'll drop 1 out of 5 words with ~40%, and we'll drop **2 out of 5 words with a probability of ~20%**. Given an utterance with just 2 words, we'll drop 1 of them with a probability of 32%. 
      *   (Note that it makes no difference whether these sentences are concatenated with something longer beforehand, as dropping words happens uniformly at random per token and hence these observations still apply).
2.  There's a mismatch between train and test time (as there are no padding tokens in the middle of test-data sentences and) because there are simply less non-zero tokens in the test-data sentences.

*   For non-sequential input like sentence features from count vectorizers this applies analogously

#### Hypotheses

*   The train-test mismatch has a bit of a negative effect.
*   (It's not too bad though because dropout is _not_ applied in exactly the same way to dense sequence features)
*   Other dropout implementations could improve results.

#### Definition of Done

Designed and documented an experimental evaluation of the current default pipeline vs....

*   [ ] disable current sparse dropout
*   [ ] regular dropout after conversion to dense embeddings
*   [ ] adapt dropout such that it scales the values (i.e. during training scale up all values of the sparse matrix to account for the dropped index-value-pairs in order to avoid a the train/test mismatch)

#### Random Follow-Up Idea:

*   [ ] adapt the dropout such that at most some constant ratio of the tokens is dropped per sentence (or none if the sentence is too short) ![](/images/icons/emoticons/help_16.png)
*   [ ] replace the zero token with a random/mask/oov kind of token instead ![](/images/icons/emoticons/help_16.png)

**Comment by alopez** [comment]
@ka-bu can you say more about the intent here? What's the problem this intended to solve, and how will we know when we've solved it (definition of done)? Add these to the description.

**Comment by ka-bu** [comment]
@alopez Sorry for the delay. Just added a description.
